bundles/testutils/tools.vitruv.testutils/src/tools/vitruv/testutils/VitruvApplicationTest.xtend

```
@@ -93,13 +92,19 @@ abstract class VitruvApplicationTest implements CorrespondenceModelContainer {
 	 * @param modelPathWithinProject A project-relative path to a model.
 	 */
 	def protected Resource resourceAt(Path modelPathWithinProject) {
-		try {
-			loadModelResource(modelPathWithinProject)
-		} catch (RuntimeException e) {
-			if (e.cause instanceof FileNotFoundException) {
-				createModelResource(modelPathWithinProject)
-			} else
-				throw e
+		synchronized (resourceSet) {
+			var Resource resource = null;
+			try {
+				resource = getAndLoadModelResource(modelPathWithinProject)
+			} catch (RuntimeException e) {
+				// EMF failed during demand creation, usually because loading from the file system failed.
+				// If it has created an empty resource, retrieve it, and otherwise create one.
+				resource = getModelResource(modelPathWithinProject)
+				if (resource === null) {
+					resource = createModelResource(modelPathWithinProject)
+				}
+			}
+			return resource
```

1. I don’t think removing the `if (e.cause instanceof FileNotFoundException)` is sensible. Other kinds of errors (say file permission errors) should be thrown. Without the check, the implementation silently (!) creates an empty `Resource` even if the target resource exists on the file system, but cannot be accessed. This is certainly not expectable behaviour, from my point of view.
2. In which case will `getModelResource` return `null` after an exception? If we don't know whether it can happen, I’d suggest doing something like `checkNotNull(getModelResource(modelPathWithinProject), "EMF did not create the Resource after not finding the source file!")` to learn whether and why it can happen.
2. Synchronizing on `resourceSet` seems unnecessarily expensive to me. Does `ResourceSet` not support loading different resources in parallel. If I understand the situation correctly, we can remove the synchronization if we remove the `createModelResource` call (see previous point). Otherwise, I’d suggest resolving the URI and synchronize on that.
2. If you want, the code could be simplified to:
```Xtend
synchronized (resourceSet) {
	return try {
		getAndLoadModelResource(modelPathWithinProject)
	} catch (RuntimeException e) {
		if (e.cause instanceof FileNotFoundException) {
			getModelResource(modelPathWithinProject) ?: createModelResource(modelPathWithinProject)
		} else throw e
	}
}
```

---

1. As I mentioned before, catching a `FileNotFoundException` is not sufficient.
2. The documentation of `getResource` says that resource creation can also throw a `RuntimeException`, thus there will be cases in which this can happen.
3. Taking a look into the EMF implementation, I would expect EMF to fail when running two `getResource` calls in parallel, especially when one resolved proxies that are contained in the other resource. In my opinion it is better to have this synchronization (since we do not perform parallel resource loadings and thus it does not produce overview) instead of failing due to concurrency issues if we, by accident, try to load resources in parallel.

---

1. Sorry, I missed your other comment. Then we should find another solution, swallowing all other errors and pretending the file does not exist, even if it might exist, is not acceptable from my point of view. This hides bugs.
2. If resource creation throws a `RuntimeException`, then `getResource` will return `null`? I’d expect the exception to be rethrown. Also: Maybe we don’t want to continue after such a `RuntimeException`?

---

1. I found the risk to miss some exception time (as we missed `ResourceException` before) to be higher and more severe than swallowing potential other exceptions, but you are right that we should find a better and finally robust solution.
2. I do not get it, but maybe it's just too late for me. The first time we try to load a resource (`getResource(..., true)`) may throw a `RuntimeException`, because the resource cannot be demand created. Then, in the `catch` block, calling `getResource(..., false)` will return `null`, as there is no resource with the given `URI` and none will be demand created and loaded (so no exception should be thrown). In that case, we create the `Resource`.
3. Maybe, yes. I was actually thinking about the Java2PCM tests, in which monitors asynchronously trigger change propagation, but they do not touch the `ResourceSet` in the test classes.

---

1. Yes, I see that you assessed the risk that way. However, I think this is a bad assessment: If we miss some exception, it will be thrown in our face, we can look at it and potentially add it to the `if`-clause. Easy. If, on the other hand, we swallow an exception (for example because of access rights), an empty resource will wrongfully be created and it might take hours to figure out what is going on (especially for somebody not familiar with the code, like us in 3 months). That’s bad!

I am insisting on this point that much because it is a pattern I have seen time and time again in the Vitruv code base: Exceptions get swallowed a lot, and the general sentiment seems to be ‘better catch too much than having another execption’. I think that that is a _fundamentally_ wrong approach.

> The first time we try to load a resource (`getResource(..., true)`) may throw a `RuntimeException`, because the resource cannot be demand created. Then, in the catch block, calling `getResource(..., false)` will return `null`, as there is no resource with the given `URI` and none will be demand created

I thought the whole point of this PR was that that is _not_ what is happening? You wrote it yourself, in the PR description: 

> `resourceSet.getResource` creates a `Resource` even if it throws an exception

---

Well, I totally agree. I think I was confused, because some proposal that we have discussed did not rethrow the exception (and would thus have also swalled other exceptions) and I think I missed that you proposed to do so in the last proposal. That is, of course, a better solution. I am just not sure yet how to properly identify the `ResourceException`, as it seems to be part of an internal API, but I will figure that out.

Regarding `getResource`: The behavior depends on the argument `loadOnDemand`, as mentioned by the documentation. Only if this is set to `true`, a `Resource` will be created and demand loaded (which fails if no persistence exists at that `URI`). If you set this to `false`, you only get the `Resource` already present in the `ResourceSet`, if existing, and otherwise `null`. My initial statement seems to be insufficient in this regard, because I did not point out that this behavior does not occur with `loadOnDemand=false`. I am sorry for that.

---

Just as a note for myself: We should apply the solution at some other places as well. Especially `ResourceRepositoryImpl` has several of these implementations, which should be fixed as well.
In `ProjectBuilderUtils`, exceptions due to these implementations are swallowed by not rethrowing them but only submitting a fatal log entry.

---

Regarding `getResource`: Yes, this is my whole point: We call `getResource` with `loadOnDemand=true` in the `try`-clause. This creates a `Resource` even if it throws an exception. Hence, when we call `getResource` with `loadOnDemand=false` in the `catch`-clause we are guaranteed to get a resource, aren’t we?

---

> This creates a Resource even if it throws an exception.
Unfortunately not. `getResource` with `loadOnDemand=true` can even fail during resource creation and then throws a `RuntimeException`. In that case, we end up in the `catch` clause and `getResource` with `loadOnDemand=false` will return `null`.

---

Yes, but will this happen if the exception’s cause is that the resource does not exist? All other exceptions should be rethrown anyway.

---

No, I would not expect that. Sorry, my answer concerned the current solution (without checking the cause of the exception) and not a solution with a cause check. In that case, simply creating the resource without trying to retrieve it should be sufficient.

There still seems to a problem: The `ResourceException`, which occurs when no file persistence of a resource with platform URI is found, is, first, internal, and, second, can also occur when some other failure during loading happened. So it is not sufficient to identify the absence of a file.
I do not really know how to properly deal with the whole situation, because investigating the exception does not seem to be a suitable solution. What if we try to load the resource, just like we do now, and if we catch an exception, we validate whether the file exists and, if not, create a resource? Otherwise, we rethrow the exception. The worst that could happen is that the file is somehow generated between the failed load and the existence check, right? This means that we may fail more often than we would do when properly investigating the exception, but to me that seems to be irrelevant.

---

> In that case, simply creating the resource without trying to retrieve it should be sufficient.

Okay. Glad that we agree.

 ---

The fact that EMF gives us no chance to see what went wrong by inspecting the exception is unbelievable. Does the `ResourceException` have a message we can recognize? I know that this solution would rely on an implementation detail, but it would still be an easy solution.

If that is not an option, I’d propose to implement the heuristic you propose, but a little more ‘clever’:

 1. Check whether the exception is a `FileNotFoundException`
 1. If not, check whether the exception’s class’ fully qualified name is that of the (not accessible) `ResourceException`.
 1. If this is the case, check whether the file exists.
 1. In all other cases, rethrow the exception.

This is all ugly, of course, but I guess it’s the best heuristic we can do.

---

The message of the `ResourceException` contains `not exist`. An implementation that works thus may look as follows:
```Xtend
try {
	return resourceSet.getResource(modelUri, true)
} catch (RuntimeException e) {
	// EMF failed during demand creation, usually because loading from the file system failed.
	// If it has created an empty resource, retrieve it, and otherwise create one.
	if (e.cause instanceof FileNotFoundException || e.message.contains("not exist")) {
		return resourceSet.createResource(modelUri)
	} else {
		throw e;
	}
}
```

Do you agree with that solution? In that case I would implement it as a central utility method and use it as a replacement for all the different implementations of resource loading that we have in the framework.

---

Yes, I agree with this one :+1:

Only thing: Please remove ‘and otherwise create one’ from the comment, since it is not accurate any more.

---

Yes, thank you! And more importantly, it should be `return resourceSet.getResource(modelUri, false)` rather than `return resourceSet.createResource(modelUri)` in the implementation, because otherwise we reimplement the original mistake ;-)

---

True :laughing:

---

This should be finally resolved by #339.